### ISSUE TITLE
DataLoader "casting" non statndard objects to lists

### ISSUE THREAD

**Issue by Arseny-N** [pytorch/pytorch#3281]
**The Problem:**
Currently if you implement a nonstandard collate function which, for example returns a namedtuple and set pin_memory=True the namedtuple would be "casted" to a list or dict by the following [code](https://github.com/pytorch/pytorch/blob/23a3f78988aeeaf3ca193ec0b03fd755ddbcdd49/torch/utils/data/dataloader.py#L116).

**Proposed Approach:**
Add another argument to the data loader constructor `DataLoader(batch_type_ctor=bt_ctor)`, where bt_ctor is a callable returning a constructed object. The result of pin_memory_batch should be passed to bt_ctor, and so should be the result of collate_fn if collate_fn is default_collate.

This approach might seem to add complexity to DataLoader since the connection between batch_type_ctor and collate_fn is non trivial if you don't read the code and may result in some code duplication.

For example if you implement a nonstandard collate_fn which returns a namedtuple then the return value of the collate_fn should be resupplied via batch_type_ctor if pin_memory is used.

But this approach seems to result in the most simple api and simple implementation.

**PS**
I would really enjoy contributing to the project.

**Comment by martinraison** [comment]
I would also like to be able to use `namedtuple` for dataset examples. This would require changes to `DataLoader` as identified above, and also to `DataParallel`. I don't know if there are other places we need to be careful (maybe `DistributedDataParallel`?)

I don't know if we want to support a `batch_type_ctor` argument, but at least handling `namedtuple` correctly seems like a good start.

Does anyone see any blockers? It's probably quite easy to implement.

**Comment by Arseny-N** [comment]
Well you can't reliably determine if an object is a namedtuple [link](https://stackoverflow.com/questions/2166818/python-how-to-check-if-an-object-is-an-instance-of-a-namedtuple) 

So either a heuristic should be used or a flag should be supplied to the DataLoader.

**Comment by martinraison** [comment]
I didn't realize that. Maybe using `batch._make([pin_memory_batch(sample) for sample in batch])` would be good enough then. We could also probably use the same trick for `DataParallel`. I know there could be false positives, but they seem rather unlikely (if necessary we could use an even stronger heuristic - checking if we're a subclass of tuple + if `_fields` exists, etc)

**Comment by Arseny-N** [comment]
Oh, cool I was not aware of the `._make` method. I agree what you are proposing is better. 

So to not duplicate code a `is_namedtuple` function should be created, where to define it - in `torch/utils/__init__.py` ?

**Comment by martinraison** [comment]
maybe the pythonic way would be to just try to call the method:

    elif isinstance(batch, collections.Sequence):
        try:
            # Handle namedtuple
            return batch._make([pin_memory_batch(sample) for sample in batch])
        except AttributeError:
            # Fallback to list
            return [pin_memory_batch(sample) for sample in batch]

An extra benefit of doing it this way is that if you really want to use your own tuple-like (but non-namedtuple) custom class, you can just add a `_make` method to it

**Comment by Arseny-N** [comment]
Good idea. I will try to setup the needed environment and write the patch on the week end.

**Comment by Arseny-N** [comment]
Hi @martinraison I've wrote the patch but fail to see how `DistributedDataParallel/DataParallel` could be affected by the changes in pin_memory. Maybe we both misunderstood each other -- my intention was to allow to write something like this
```python
def collate(batch):
  data, labels, dataset_ids = zip(*batch)
  return Batch(data=torch.cat(data), labels=labels, ids=dataset_ids)

validate_ds = DataLoader(dataset, collate_fn=collate)
...
for batch in validate_ds:
  output = model(batch.data)
  ...
  print(f'ids {batch.ids} Predicted {output} Labels {batch.labels}')
  ...
```
-------
Both  `DistributedDataParallel` and `DataParallel` operate on the `model` so the `namedtuple` returned by the `collate_fn` does not affect them, unless it is passed as an argument to the `model`. Am I overlooking something ?
**PS** Sorry it took so long, had some technical problems in setting up Pytorch in build mode.

**Comment by martinraison** [comment]
Hi @Arseny-N, indeed, my idea was that you should be able to pass a namedtuple as an argument to the model directly. Here you only pass a single tensor so you can just do `batch.data`, but sometimes you have a bunch of different tensors as inputs, and it's useful to pass everything to the model inside a namedtuple. This being said, we don't have to fix everything at once.

**Comment by sergiogiro** [comment]
Hi, I faced a related issue so I'll try to discuss here before raising a new issue (as I think that there's a solution that might fix both issues).

Pinning memory with the current implementation might cause an unnecessary slowdown in performance. The root cause is that, when the dataloader pins a batch into memory, it calls the pin_memory_batch function, that iterates through the batch structure searching for tensors "as possible" (as long as elements are sequences or mappings). This search can create a lot of tight loops, in my case with penalties in performance of around 2x slower. In my particular case, I'm loading data from files and, in addition to the tensors, in the batch object I have lists of (non-numeric) data I use for post-processing. I don't think this scenario is very unusual. These lists contain one element for each value in the tensors, so it seems a lot of time is spent going through these massive lists and testing whether each value is a tensor or not. I can make the issue go away by wrapping the lists (or any data that is explored recursively) inside a holder, that is simply a class that is opaque to the search (ie, neither a sequence nor a mapping). The solution is not very elegant though.

I was wondering if, in the same way that you can specify a collate function, we could implement an optional get_tensors_to_pin function, that given a batch would return, well, the elements that you want in pinned memory. This would work also for namedtuples (even if they're immutable, the tensor would be the same, it's just that .pin_memory() would be called on it). There'll be no need to "reconstruct" the batch object afterwards, as it wouldn't need to be deconstructed in the first place.

This is my first "contribution" to pytorch. Do you think it's a feasible solution? Do you think it deserves a new issue of its own?

Best,
Sergio
